### PR TITLE
chore: migrate tests to TypeScript.

### PR DESCRIPTION
**Pull request by jackfranklin**
This CL migrates all the tests to TypeScript. The main benefits of this
is that we start consuming our TypeScript definitions and therefore find
errors in them. The act of migrating found some bugs in our definitions
and now we can be sure to avoid them going forwards.

You'll notice the addition of some `TODO`s in the code; I didn't want
this CL to get any bigger than it already is but I intend to follow
those up once this lands. It's mostly figuring out how to extend the
`expect` types with our `toBeGolden` helpers and some other slight
confusions with types that the tests exposed.

### CODE REVIEW

File: test/click.spec.ts [hunk 1]
```diff
@@ -30,7 +31,7 @@ describe('Page.click', function () {
 
     await page.goto(server.PREFIX + '/input/button.html');
     await page.click('button');
-    expect(await page.evaluate(() => result)).toBe('Clicked');
+    expect(await page.evaluate(() => globalThis.result)).toBe('Clicked');
```
Comment (jackfranklin) [inline test/click.spec.ts:34]: this is to keep TypeScript happy - it's happy for you to do `globalThis.X` but not just `X` or `window.X`.

Comment (mathiasbynens) [reply test/click.spec.ts:34]: Ack. I actually think this is better since it's more explicit.

**Review by mathiasbynens — APPROVED** [review]

File: tsconfig.json [hunk 2]
```diff
@@ -8,9 +8,10 @@
     "module": "CommonJS",
     "declaration": true,
     "declarationMap": true,
-    "esModuleInterop": true
+    "esModuleInterop": true,
+    "resolveJsonModule": true
   },
   "include": [
-    "src"
+    "src",
```
Comment (mathiasbynens) [inline tsconfig.json]: This is invalid JSON, but I guess tsc doesn't care?

Comment (mathiasbynens) [inline tsconfig.json]:
```suggestion
    "src"
```

File: test/assets/input/textarea.html [hunk 3]
```diff
@@ -8,8 +8,8 @@
     <script src='mouse-helper.js'></script>
     <script>
       window.result = '';
-      let textarea = document.querySelector('textarea');
+      window.textarea = document.querySelector('textarea');
```
Comment (mathiasbynens) [inline test/assets/input/textarea.html]: Might as well use `globalThis` here too, for consistency

File: test/assets/input/textarea.html [hunk 4]
```diff
@@ -8,8 +8,8 @@
     <script src='mouse-helper.js'></script>
     <script>
       window.result = '';
```
Comment (mathiasbynens) [inline test/assets/input/textarea.html]:
```suggestion
      globalThis.result = '';
```

**Pull request merged**